# TUGAS PRAKTIKUM 2 MACHINE LEARNING #

Mochammad Dzahwan Fadhloly || 24060121140168 || Lab ML A1

# Nomor 1

Dataset Bunga Iris

Link: http://archive.ics.uci.edu/dataset/53/iris

**Import Library yang Dibutuhkan**

Pertama gunakan library pandas. Pandas merupakan sebuah library di Python yang berlisensi BSD dan open source yang menyediakan struktur data dan analisis data yang mudah digunakan.

In [ ]:
#import library yang dibutuhkan
import pandas

**Memuat Dataset**

Pada tugas ini dataset yang digunakan adalah dataset klasifikasi Bunga Iris.

Pertama definisikan variabel names untuk menyimpan nama-nama kolom atau atribut dari dataset. Setiap elemen dalam daftar names mewakili nama kolom yang ada dalam dataset.

Kemudian digunakan library pandas untuk membaca dataset dari file CSV yang sudah diimport tadi yaitu dengan fungsi read_csv dan memasukkan dua argumen untuk dibaca yaitu file csv 'iris.data' dan names yang sudah didefinisikan tadi. Maka dataset yang digunakan sudah berbentuk DataFrame.

In [ ]:
#mendefinisikan list berisi kolom-kolom dataset
names = ['sepal.lenght','sepal-width','petal-lenght','petal-width','class']

#membaca dataset dengan library pandas
dataset = pandas.read_csv('iris.data', names=names)

Kemudian kode dibawah digunakan untuk menampilkan dimensi dari dataset yang telah dibaca menggunakan pandas. Dimensi dari dataset merupakan gambaran singkat mengenai banyaknya jumlah baris yang menunjukkan banyaknya sampel data dan jumlah kolom yang menunjukkan atribut data dari dataset terkait.

In [ ]:
# menentukan dimensi dari dataset
print(dataset.shape)

(150, 5)


Hasil/output yang dihasilkan adalah (150, 5) yang berarti dataset memiliki 150 baris dan 5 kolom.

Selanjutnya mencetak dataset 10 baris pertama yaitu dengan menggunakkan head untuk melihat sebagian awal dari dataset dan memahami isi data secara lebih rinci. Dapat dilihat output yang dikeluarkan adalah 10 data pertama dalam dataset.

In [ ]:
# menampilkan 10 data teratas dari dataset
print(dataset.head(10))

   sepal.lenght  sepal-width  petal-lenght  petal-width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa
5           5.4          3.9           1.7          0.4  Iris-setosa
6           4.6          3.4           1.4          0.3  Iris-setosa
7           5.0          3.4           1.5          0.2  Iris-setosa
8           4.4          2.9           1.4          0.2  Iris-setosa
9           4.9          3.1           1.5          0.1  Iris-setosa


**I. Membuat Validasi Dataset**

Validasi dilakukan untuk mengetahui bahwa model yang dibuat bagus. Disini, kita membagi datatest yang telah dimuat menjadi dua, 80% diantaranya akan digunakan untuk melatih model dan 20% digunakan untuk data validasi. (Dataset yang digunakan berupa 150 bunga Iris pada praktikum sebelumnya)

In [ ]:
#import library yang dibutuhkan
import sklearn
from sklearn import model_selection

#menyimpan nilai-nilai pada dataset ke dalam variabel array
array = dataset.values

#menyimpan data fitur dataset pada variabel X
X = array[:,0:4]

#menyimpan data fitur dataset pada variabel Y
Y = array[:,4]

#mendefinisikan ukuran testing data dan seed
validation_size = 0.20
seed = 7

#memisahkan data menjadi training dan testing
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

In [ ]:
#mengecek data yang sudah dibagi
X_train.shape

(120, 4)

Setelah perintah diatas dieksekusi, kita sudah memiliki dua data, yaitu X_train dan Y_train untuk mempersiapkan model dan rangkaian X_validation dan Y_validation ynag dapat digunakan selanjutnya.

**II. K-Folds Cross Validation**

Kita akan menggunakan validasi silang 10 kali lipat untuk memperkirakan akurasi. Untuk itu dataset dibagi menjadi 10 bagian, 9 untuk latihan dan 1 untuk pengujian dan ulangi untuk semua kombinasi.

In [ ]:
#test options and evaluation metric
seed = 7
scoring = 'accuracy'

**III. Membangun Model**

Untuk mengetahui algoritma yang cocok dengan studi kasus ini maka kita harus mengevaluasi dengan beberapa algoritma. Pada tugas praktikum kedua ini, digunakan tiga model yaitu Stochastic Gradient Descent (SGD), Decission Tree, dan Bagging Classifier.

In [ ]:
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier

# Spot Check Algorithms
models = []
models.append(('SGD', SGDClassifier()))
models.append(('DT', DecisionTreeClassifier()))
models.append(('BC', BaggingClassifier()))

# evaluate each model in turn
results = []
names = []
for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

SGD: 0.800000 (0.130171)
DT: 0.950000 (0.076376)
BC: 0.966667 (0.040825)


**IV. Memilih Model Terbaik**

Jika sudah memiliki hasil evaluasi dari ketiga model diatas untuk memilih model terbaik dilakukan dari perbandingan satu sama lainnya dan dipilih yang paling akurat. Dari eksekusi script diatas, kita mendapatkan hasil mentah sebagai berikut:
1. SGD: 0.800000 (0.130171)
2. DT: 0.950000 (0.076376)
3. BC: 0.966667 (0.040825)

Dari hasil di atas, kita dapat melihat bahwa Bagging Classifier memiliki nilai akurasi perkiraan terbesar. Setelah mengetahui model yang paling akurat yaitu Bagging Classifier (BC), selanjutnya kita dapat mencoba melakukan pengujian tentang keakuratan model Bagging Classifier terhadap data yang ada.

Pertama impor fungsi-fungsi yang diperlukan yang berguna untuk mengevaluasi performa model klasifikasi. Kemudian model gaus (BaggingClassifier) dilatih dengan menggunakan data X_train dan Y_train.
Setelah itu cetak nilai akurasi dengan fungsi accuracy_score yang merupakan metrik yang mengukur sejauh mana model klasifikasi berhasil memprediksi label. Kemudian Confussion matrix adalah tabel yang digunakan untuk menggambarkan kinerja model klasifikasi dan classification_report akan mencetak laporan yang berisi precision, recall, f1-score, dan support.

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

# Make predictions on validation dataset
gaus = BaggingClassifier()

gaus.fit(X_train, Y_train)
predictions = gaus.predict(X_validation)

print('Akurasi pada Data Testing : ',accuracy_score(Y_validation, predictions))
print('Confusion Matrix : \n',confusion_matrix(Y_validation, predictions))
print('\n Classification Report : \n',classification_report(Y_validation, predictions))

Akurasi pada Data Testing :  0.9
Confusion Matrix : 
 [[ 7  0  0]
 [ 0 10  2]
 [ 0  1 10]]

 Classification Report : 
                  precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.91      0.83      0.87        12
 Iris-virginica       0.83      0.91      0.87        11

       accuracy                           0.90        30
      macro avg       0.91      0.91      0.91        30
   weighted avg       0.90      0.90      0.90        30



**Kesimpulan**

Dari hasil pengujian diatas, untuk dataset Bunga Iris didapatkan nilai accuracy pada data testing adalah 0.9

# Nomor 2
Dataset Wine

Link: http://archive.ics.uci.edu/dataset/109/wine

**Import Library yang Dibutuhkan**

Pertama gunakan library pandas. Pandas merupakan sebuah library di Python yang berlisensi BSD dan open source yang menyediakan struktur data dan analisis data yang mudah digunakan.

In [ ]:
#import library yang dibutuhkan
import pandas

**Memuat Dataset**

Pada tugas ini dataset yang digunakan adalah dataset klasifikasi Wine.

Pertama definisikan variabel names untuk menyimpan nama-nama kolom atau atribut dari dataset. Setiap elemen dalam daftar names mewakili nama kolom yang ada dalam dataset.

Kemudian digunakan library pandas untuk membaca dataset dari file CSV yang sudah diimport tadi yaitu dengan fungsi read_csv dan memasukkan dua argumen untuk dibaca yaitu file csv 'wine.data' dan names yang sudah didefinisikan tadi. Maka dataset yang digunakan sudah berbentuk DataFrame.

In [ ]:
#mendefinisikan list berisi kolom-kolom dataset
names = ['class', 'Alcohol', 'Malicacid', 'Ash', 'Alcalinity_of_ash', 'Magnesium', 'Total_phenols', 'Flavanoids', 'Nonflavanoid_phenols', 'Proanthocyanins']
#membaca dataset dengan library pandas
dataset = pandas.read_csv('wine.data', names=names)

Kemudian kode dibawah digunakan untuk menampilkan dimensi dari dataset yang telah dibaca menggunakan pandas. Dimensi dari dataset merupakan gambaran singkat mengenai banyaknya jumlah baris yang menunjukkan banyaknya sampel data dan jumlah kolom yang menunjukkan atribut data dari dataset terkait.

In [ ]:
#menentukan dimensi dari dataset
print(dataset.shape)

(178, 10)


Hasil/output yang dihasilkan adalah (178, 10) yang berarti dataset memiliki 178 baris dan 10 kolom.

Selanjutnya mencetak dataset 10 baris pertama yaitu dengan menggunakkan head untuk melihat sebagian awal dari dataset dan memahami isi data secara lebih rinci. Dapat dilihat output yang dikeluarkan adalah 10 data pertama dalam dataset.

In [ ]:
#menampilkan 10 data teratas dari dataset
print(dataset.head(10))

                   class  Alcohol  Malicacid   Ash  Alcalinity_of_ash  \
1 14.23 1.71 2.43   15.6      127       2.80  3.06               0.28   
  13.20 1.78 2.14   11.2      100       2.65  2.76               0.26   
  13.16 2.36 2.67   18.6      101       2.80  3.24               0.30   
  14.37 1.95 2.50   16.8      113       3.85  3.49               0.24   
  13.24 2.59 2.87   21.0      118       2.80  2.69               0.39   
  14.20 1.76 2.45   15.2      112       3.27  3.39               0.34   
  14.39 1.87 2.45   14.6       96       2.50  2.52               0.30   
  14.06 2.15 2.61   17.6      121       2.60  2.51               0.31   
  14.83 1.64 2.17   14.0       97       2.80  2.98               0.29   
  13.86 1.35 2.27   16.0       98       2.98  3.15               0.22   

                   Magnesium  Total_phenols  Flavanoids  Nonflavanoid_phenols  \
1 14.23 1.71 2.43       2.29           5.64        1.04                  3.92   
  13.20 1.78 2.14       1.28      

**I. Membuat Validasi Dataset**

Validasi dilakukan untuk mengetahui bahwa model yang dibuat bagus. Disini, kita membagi datatest yang telah dimuat menjadi dua, 80% diantaranya akan digunakan untuk melatih model dan 20% digunakan untuk data validasi. (Dataset yang digunakan berupa 178 data Wine)

In [ ]:
#import library yang dibutuhkan
import sklearn
from sklearn import model_selection

#menyimpan nilai-nilai pada dataset ke dalam variabel array
array = dataset.values

X = array[:,0:9]

Y = array[:,1]

validation_size = 0.20
seed = 7

#memisahkan data menjadi training dan testing
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

#mengecek data yang sudah dibagi
X_train.shape

(142, 9)

Setelah perintah diatas dieksekusi, kita sudah memiliki dua data, yaitu X_train dan Y_train untuk mempersiapkan model dan rangkaian X_validation dan Y_validation ynag dapat digunakan selanjutnya.

**II. K-Folds Cross Validation**

Kita akan menggunakan validasi silang 10 kali lipat untuk memperkirakan akurasi. Untuk itu dataset dibagi menjadi 10 bagian, 9 untuk latihan dan 1 untuk pengujian dan ulangi untuk semua kombinasi.

In [ ]:
#test options and evaluation metric
seed = 7
scoring = 'accuracy'

**III. Membangun Model**

Untuk mengetahui algoritma yang cocok dengan studi kasus ini maka kita harus mengevaluasi dengan beberapa algoritma. Pada tugas praktikum kedua ini, digunakan tiga model yaitu K-Nearest Neughbors (KNN), Gaussian Naive Bayes (NB), dan Support Vector Machines (SVM).

In [ ]:
# import library yang dibutuhkan
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

#spot check algorithms
models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

#evaluate each model in turn
results = []
names = []
for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.289524 (0.122647)
NB: 0.704762 (0.073925)
SVM: 0.050000 (0.055787)


**IV. Memilih Model Terbaik**

Jika sudah memiliki hasil evaluasi dari ketiga model diatas untuk memilih model terbaik dilakukan dari perbandingan satu sama lainnya dan dipilih yang paling akurat. Dari eksekusi script diatas, kita mendapatkan hasil mentah sebagai berikut:
1. KNN: 0.289524 (0.122647)
2. NB: 0.704762 (0.073925)
3. SVM: 0.050000 (0.055787)

Dari hasil di atas, kita dapat melihat bahwa Gaussian Naive Bayes (NB) memiliki nilai akurasi perkiraan terbesar. Setelah mengetahui model yang paling akurat yaitu Gaussian Naive Bayes (NB), selanjutnya kita dapat mencoba melakukan pengujian tentang keakuratan model Gaussian Naive Bayes terhadap data yang ada.

Pertama impor fungsi-fungsi yang diperlukan yang berguna untuk mengevaluasi performa model klasifikasi. Kemudian model nb (GaussianNB) dilatih dengan menggunakan data X_train dan Y_train.
Setelah itu cetak nilai akurasi dengan fungsi accuracy_score yang merupakan metrik yang mengukur sejauh mana model klasifikasi berhasil memprediksi label. Kemudian Confussion matrix adalah tabel yang digunakan untuk menggambarkan kinerja model klasifikasi dan classification_report akan mencetak laporan yang berisi precision, recall, f1-score, dan support.

In [ ]:
# import library yang dibutuhkan
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# make predictions on validation dataset
nb = GaussianNB()
nb.fit(X_train, Y_train)
predictions = nb.predict(X_validation)
print('Akurasi pada Data Testing : ', accuracy_score(Y_validation, predictions))
print('Confusion Matrix : \n',confusion_matrix(Y_validation, predictions))
print('\n Classification Report : \n',classification_report(Y_validation, predictions, zero_division=1))

Akurasi pada Data Testing :  0.8611111111111112
Confusion Matrix : 
 [[1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 

**Kesimpulan**

Dari hasil pengujian di atas, untuk dataset Wine didapatkan nilai accuracy pada data testing adalah 0.8611111111111112